netty/src/main/java/io/grpc/netty/NettyClientTransport.java

```diff
@@ -199,6 +210,32 @@ public void operationComplete(ChannelFuture future) throws Exception {
             Status.INTERNAL.withDescription("Connection closed with unknown cause"));
       }
     });
+
+    if (deadline != null) {
+      Runnable onTimeout = new Runnable() {
+        @Override
+        public void run() {
+          if (connectFuture instanceof Promise) {
+            try {
+              ((Promise<?>) connectFuture)
+                  .tryFailure(new SocketTimeoutException("connection timeout"));
+              return;
+            } catch (IllegalStateException e) {
+              // ignore and continue
+            }
+          }
+          if (writeFuture instanceof Promise) {
+            try {
+              ((Promise<?>) writeFuture)
+                  .tryFailure(new ConnectTimeoutException("write timeout"));
+            } catch (IllegalStateException e) {
+              // ignore
+            }
+          }
+        }
+      };
+      handler.startDeadlineWatchdog(onTimeout, deadline);
```

Comment:
Instead of doing this on this thread, you can do it on the event loop. Instead of passing `negotiationHandler` to `Bootstrap` you can pass an instance of something like:
```java
class MyInit extends ChannelInitializer {
  protected void initChannel(C ch) throws Exception {
    handler.startDeadlineWatchdog(onTimeout, deadline);
    channel.pipeline().addFirst(negotiationHandler);
  }
}
```

At that point, all watchdog operations are happening on the event loop and you don't need the lock in NettyClientHandler.